controllers/finance_controller.py

```diff
@@ -1,26 +1,103 @@
-from tracker.finance_tracker import FinanceTracker
 from datetime import datetime
 
+from sqlalchemy import func, extract, and_
+from sqlalchemy.exc import IntegrityError
 
-class FinanceController:
+from database import SessionLocal
+from models.expense import Expense
+from models.investment import Investment
+from views.utilities import validate_date, validate_and_convert_date
+
+
+class FinanceTracker:
     def __init__(self):
-        self.finance_tracker = FinanceTracker()
+        # Initialize a new SQLAlchemy session
+        self.db_session = SessionLocal()
 
-    def add_expense(self, user_id, category, amount, date_str):
-        success, message = self.finance_tracker.add_expense(user_id, category, amount, date_str)
-        return success, message
+    def add_expense(self, user_id, category, amount, date):
+        # First, validate the date format
+        date_obj, message = validate_and_convert_date(date)
+        if date_obj is None:
+            return False, message  # Return early if the date is invalid
+        try:
+            date_obj = datetime.strptime(date, "%Y-%m-%d").date()
+            new_expense = Expense(user_id=user_id, category=category, amount=amount, date=date_obj)
+            self.db_session.add(new_expense)
+            self.db_session.commit()
+            return True, "Expense added successfully."
+        except IntegrityError as e:
+            self.db_session.rollback()  # Rollback in case of an error
+            return False, "Failed to add expense: a similar expense already exists."
+        except Exception as e:
+            self.db_session.rollback()  # Rollback the session to undo the operation in case of error
+            return False, f"Failed to add expense: {e}"
 
-    def add_investment(self, user_id, investment_type, amount, date_str, returns=None):
-        success, message = self.finance_tracker.add_investment(user_id, investment_type, amount, date_str, returns)
-        return success, message
+    def add_investment(self, user_id, type, amount, date, returns=None):
```

Comment:
same for investment, make a data representation class, fill it, then pass it. do not make add() methods to CREATE the obj & then save it.
just to save it, single responsibility.

Beyond I feel 

class FinanceRecord could be the obj e.g.
user_id, category, amount, date (expense)
user_id, type, amount, date (investment)

```
unto FinanceRecord
user_id
type
category
amount
date
```

Reply:
Same as previous